### DISCUSSION TITLE
Create Markdown Page Via GraphQL

### DISCUSSION THREAD

**Discussion by CodeCook83** [requarks/wiki#6314] [Help / Questions]
Hi and thanks for the awesome wiki

I'm trying to create a page via GraphQL. 

When I send the content of the page as markdown instructions, the page is created, but the markdown instructions are not visualized. They are simply displayed as e.g. `# Header`. If the page was previously created as a markdown page and I just update the content via GraphQL with the same markdown instructions, the content is displayed correctly. What am I doing wrong? Is there an option in the mutation that I missed?

Here is the code for creating.
```javascript
const CERATE_WIKI_PAGE_MUTATION = gql`
  mutation Page {
    pages {
      create(
        title: "${title}"
        content: "${content}"
        description: "${description}"
        editor: "${editor}"
        isPublished: ${isPublished}
        isPrivate: ${isPrivate}
        locale: "${locale}"
        path: "${path}"
        tags: "${tags}"
      ) {
        responseResult {
          succeeded
          errorCode
          slug
          message
        }
        page {
          id
          path
          title
        }
      }
    }
  }
`;
const result = await this.client.mutate({
	mutation: CERATE_WIKI_PAGE_MUTATION,
	variables: {
	  title,
	  content,
	  description,
	  editor,
	  isPublished,
	  isPrivate,
	  locale,
	  path,
	  tags,
	},
});
```

Here is the code for updating.
```javascript
const UPDATE_WIKI_PAGE_MUTATION = gql`
	mutation update {
	  pages {
	    update(
	      id: ${id}
	      title: "${title}"
	      content: "${content}"
	      description: "${description}"
	      editor: "${editor}"
	      isPublished: ${isPublished}
	      isPrivate: ${isPrivate}
	      locale: "${locale}"
	      tags: "${tags}"
	    ) {
	      responseResult {
	        succeeded
	        errorCode
	        slug
	        message
	      }
	      page {
	        id
	        content
	        title
	      }
	    }
	  }
	}
`;
const result = await this.client.mutate({
	mutation: UPDATE_WIKI_PAGE_MUTATION,
	variables: {
	  id,
	  title,
	  content,
	  description,
	  editor,
	  isPublished,
	  isPrivate,
	  locale,
	  tags,
	},
});
```

However, this does not apply if I send HTML instructions in the content. They visualize correctly when creating or updating. But if I send HTML instructions as content (like the example below) , I get the error message `Syntax Error: Unterminated string`, if I send it without line breaks, it works fine. Although this is not my problem. It was very time consuming to debug and fix this problem. If you give me a solution to this as well, it will be great.

Bad HTML example
```javascript
const content = `<h1>List</h1>
    <table>
      <tr>
        <th>Name</th>
        <th>Mail</th>
        <th>Phone Number</th>
      </tr>
      <tr>
        <td>John Doe</td>
        <td>example@domain.com</td>
        <td>12345678</td>
      </tr>
    </table>`;
```
Correct HTML example
```javascript
const content = `<h1>List</h1><table><tr><th>Name</th><th>Mail</th><th>Phone Number</th></tr><tr><td>John Doe</td><td>example@domain.com</td><td>12345678</td></tr></table>`;
```
I have this versions
```
OS: Ubuntu 22.04.2 and Docker
Wiki.js version: 2.5.268
Database engine: PostgreSQL 11.1.4
```

In my code, I'm using 
```
macOS 12.6.4
Node.js 18.12.1
@apollo/client 3.7.10
```

**Comment by CodeCook83** [comment]
I found a variable called `contentType` in the server code that specifies the context. This can take three values: `markdown`, `html` and `json`. 

When I send this variable with the createPage function, I get the following error message: `ApolloError: Response not successful: Received status code 400` (Bad request).

In the `page.graphql` file there is `contentType` neither in create nor in update. Is this the reason?

**Reply by E314c** [reply]
I was just messing around with this myself and the value you actually want to be setting is the `editor` and saying it's `markdown`.

So for creating a new page (or updating an existing one) where the `content` you submit is in markdown, you'd submit:

```graphql
mutation {
	pages {
    create(
      title: "API markdown test page"
      content:"# Header\n\n```js\nconst x = 5;\n```"
      description: "Page made through API"
      editor: "markdown"
      isPublished:true
      isPrivate: false
      locale: "en"
      path: "/api-markdown-test-page"
      tags:[ "api", "test"]
    ){
      responseResult {
        message
      }
      page {
        id
        path
        contentType
      }
    }
  }
}
```


If I were to speculate, the `editor` input argument represents what type of frontend editor you used to create the content, though I would like to see it given an `enum` schema value or description as I initially thought it was used as a synonym for "author".

From some playing around (snooping on the requests sent when modifying various pages in different editors) the available values include:
- `ckeditor` when using the "Visual Editor", aka "What You See Is What You Get" (wysiwyg) editor
- `code` for the html editor
- `asciidoc` for the "AsciiDoc editor"
- `markdown` for the "Markdown editor"
